packages/studio-base/src/panels/ThreeDeeRender/SceneExtensionConfig.ts

```diff
@@ -0,0 +1,122 @@
+
+import { Cameras } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/Cameras";
+import { FoxgloveGrid } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/FoxgloveGrid";
+import { FrameAxes } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/FrameAxes";
+import { Grids } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/Grids";
+import { ImageMode } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/ImageMode/ImageMode";
+import { Images } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/Images";
+import { LaserScans } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/LaserScans";
+import { Markers } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/Markers";
+import { OccupancyGrids } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/OccupancyGrids";
+import { PointClouds } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/PointClouds";
+import { Polygons } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/Polygons";
+import { PoseArrays } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/PoseArrays";
+import { Poses } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/Poses";
+import { PublishSettings } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/PublishSettings";
+import { FoxgloveSceneEntities } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/SceneEntities";
+import { SceneSettings } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/SceneSettings";
+import { Urdfs } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/Urdfs";
+import { VelodyneScans } from "@foxglove/studio-base/panels/ThreeDeeRender/renderables/VelodyneScans";
+
+import { IRenderer } from "./IRenderer";
+import { SceneExtension } from "./SceneExtension";
+import { MeasurementTool } from "./renderables/MeasurementTool";
+import { PublishClickTool } from "./renderables/PublishClickTool";
+import { InterfaceMode } from "./types";
+
+export type SceneExtensionConfig = {
+  reserved: ReservedSceneExtensionConfig;
+  extensionsById: Record<AvailableNames, ExtensionOverride<SceneExtension>>;
+};
+/** Reserved because the Renderer has members that reference them specifically */
+
+export type ReservedSceneExtensionConfig = {
+  imageMode: ExtensionOverride<ImageMode>;
+  measurementTool: ExtensionOverride<MeasurementTool>;
+  publishClickTool: ExtensionOverride<PublishClickTool>;
+  //   cameraHandler: ExtensionOverride<ICameraHandler>;
+};
+
+export type ExtensionOverride<ExtensionType extends SceneExtension> = {
+  init: (renderer: IRenderer) => ExtensionType;
+  /** Which interfaceModes this extension is supported in. If undefined, will default to both '3d' and 'image' modes */
+  supportedModes?: InterfaceMode[];
+};
+
+type ReservedNames = keyof ReservedSceneExtensionConfig;
+type AvailableNames = Exclude<string, ReservedNames>;
```

Comment:
You can't actually exclude things from `string` like that, I think Exclude only works for unions:
```suggestion
type AvailableNames = Exclude<string, ReservedNames>;
```

Reply:
Really had me going when it didn't error :(